File: apps/web/src/lib/referencePosition.ts
// TypeScript twin of `position_vs_source_interval`
// (apps/api/src/hadp_api/modules/observations/reference_position.py). It states only a
// deterministic FACT about the value versus the LAB-provided interval — never an introduced
// "optimal"/normal bound, never a verdict. Keep this truth table in sync with the Python twin;
// the authoritative cases live in apps/api/tests/test_reference_position.py.
//
// When the 0009 out_of_source_interval rule lands, this presenter-side computation is replaced by
// a read of the recorded RuleEvaluation — same positions, same copy keys, so it is a data-source
// swap, not a rewrite.

export type ReferencePosition =
  | "above"
  | "below"
  | "within"
  | "no_reference"
  | "not_evaluable";

export interface PositionBar {
  hasScale: boolean;
  bandStartPct: number;
  bandEndPct: number;
  midPct: number; // band midpoint — the reference tick + "Referenz x–y" label sit here
  dotPct: number | null;
  scaleMax: number; // right-hand scale-end value (left end is always 0)
}

function num(v: string | null | undefined): number | null {
  if (v == null || v === "") return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

/** Position of `value` relative to the lab interval `[low, high]`. Mirrors the Python twin. */
export function referencePosition(
  value: string | null | undefined,
  referenceLow: string | null | undefined,
  referenceHigh: string | null | undefined,
): ReferencePosition {
  const v = num(value);
  const lo = num(referenceLow);
  const hi = num(referenceHigh);
  if (v == null) return "not_evaluable";
  if (lo == null && hi == null) return "no_reference";
  if (hi != null && v > hi) return "above";
  if (lo != null && v < lo) return "below";
  if (lo != null && hi != null) return "within";
  return "not_evaluable";
}

// Position bar geometry (the Claude Design "Lage zur Referenz" comp). Scale is [0, scaleMax] —
// natural for lab values (≥ 0). scaleMax fits the larger of the upper reference bound and the value
// (+ a little headroom) so the value dot is always on-scale and "how far outside" stays readable.
// The reference BAND reflects ONLY the lab-provided interval — never an introduced "optimal"/normal
// bound, never a clinical target; the axis extent is a cosmetic choice, not a bound. The bar is
// decorative: the pill + screen-reader sentence carry the meaning (never colour alone).
export function referenceBar(
  value: string | null | undefined,
  referenceLow: string | null | undefined,
  referenceHigh: string | null | undefined,
): PositionBar {
  const empty: PositionBar = {
    hasScale: false,
    bandStartPct: 0,
    bandEndPct: 0,
    midPct: 0,
    dotPct: null,
    scaleMax: 0,
  };
  const v = num(value);
  const lo = num(referenceLow);
  const hi = num(referenceHigh);
  // A two-sided, non-degenerate interval is required to draw a meaningful band.
  if (lo == null || hi == null || hi <= lo) return empty;

  const top = Math.max(hi, v ?? hi);
  const scaleMax = top * 1.07; // headroom so the dot/end never sits flush against the edge
  const pct = (x: number): number => {
    const p = (x / scaleMax) * 100;
    return p < 0 ? 0 : p > 100 ? 100 : p;
  };

  return {
    hasScale: true,
    bandStartPct: pct(lo),
    bandEndPct: pct(hi),
    midPct: pct((lo + hi) / 2),
    dotPct: v == null ? null : pct(v),
    scaleMax,
  };
}

File: apps/api/src/hadp_api/modules/observations/reference_position.py
"""Pure predicate: where a value sits relative to its lab-provided reference interval.

This is the SINGLE source of truth for the above/below/within determination. It states
only a deterministic FACT about the value versus the laboratory-provided interval — it
never introduces an "optimal"/normal bound and never asserts a clinical verdict.

Two callers rely on these exact semantics: the deterministic narrative provider
(`modules/reports/narrative.py`) and the web "Lage zum Referenzintervall" column. The
TypeScript twin in `apps/web/src/lib/referencePosition.ts` mirrors this truth table; the
authoritative cases live in `apps/api/tests/test_reference_position.py`. The planned
`out_of_source_interval` rule (docs/notes/0009) reuses this helper so the comparison has
one implementation, not three (the drift the 0009 review flagged).
"""

from __future__ import annotations

from decimal import Decimal
from typing import Literal

ReferencePosition = Literal["above", "below", "within", "no_reference", "not_evaluable"]


def position_vs_source_interval(
    value: Decimal | None,
    reference_low: Decimal | None,
    reference_high: Decimal | None,
) -> ReferencePosition:
    """Position of ``value`` relative to the lab interval ``[low, high]``.

    - no value                            -> ``"not_evaluable"``
    - no bound at all                     -> ``"no_reference"``
    - value strictly above the high bound -> ``"above"``
    - value strictly below the low bound  -> ``"below"``
    - both bounds present, not exceeded   -> ``"within"``
    - one-sided bound, not exceeded       -> ``"not_evaluable"`` (cannot place in an interval)

    Equality with a bound is ``"within"`` (strict inequalities), matching the narrative
    provider. A one-sided interval never yields a guessed ``"within"``.
    """
    if value is None:
        return "not_evaluable"
    if reference_low is None and reference_high is None:
        return "no_reference"
    if reference_high is not None and value > reference_high:
        return "above"
    if reference_low is not None and value < reference_low:
        return "below"
    if reference_low is not None and reference_high is not None:
        return "within"
    return "not_evaluable"
